qcodes_contrib_drivers/drivers/NationalInstruments/RFSG.py

```diff
@@ -0,0 +1,245 @@
+                                           NIRFSG_ATTR_REF_CLOCK_SOURCE),
+                           set_cmd=self.set_clock_source,
+                           )
+
+        self.initiate()
+        self.connect_message()
+
+    def initiate(self):
+        """
+        Initiate signal generation. This causes the NI-RFSG device to leave
+        the Configuration state.
+        """
+        self.wrapper.Initiate(self._handle)
+
+    def abort(self):
+        """
+        Stop signal generation and return to the Configuration state.
+        """
+        self.wrapper.Abort(self._handle)
+
+    def ConfigureRF(self, frequency: float, power_level: float,
+                    initiate: bool):
+        """
+        NI-RFSG devices can only set both the frequency and power level
+        simultatneously. Convenience methods are defined below for setting
+        them individually.
+
+        NOTE: PXI-5670/5671 and PXIe-5672 devices must be in the Configuration
+        state before calling this function (by calling abort()), that is not
+        checked here.
+
+        Args:
+            frequency: frequency in Hz
+            power_level: power level in dBm
+            initiate: if True, call self.initiate after configuring, which
+                starts RF output
+        """
+        self.wrapper.ConfigureRF(self._handle, ViReal64(frequency),
+                                 ViReal64(power_level))
+        if initiate:
+            self.initiate()
+
+    def set_frequency(self, frequency: float, initiate: bool = False):
+        power_level = self.get_attribute(NIRFSG_ATTR_POWER_LEVEL)
+        self.ConfigureRF(frequency, power_level, initiate)
+
+    def set_power_level(self, power_level: float, initiate: bool = False):
+        frequency = self.get_attribute(NIRFSG_ATTR_FREQUENCY)
+        self.ConfigureRF(frequency, power_level, initiate)
+
+    def set_clock_source(self, source: str, clock_rate: float = 10e6):
+        """
+        Sets reference clock source
+        Args:
+            source: reference clock source. Valid values are
+                        onboard
+                        clk_in
+                        ref_in
+                        pxi_clk
+                        ref_in_2
+                        pxi_clk_master
+                    Last two values are valid for PXIe-5840 with PXIe-5653.
+            clock_rate: Not implemented. Reference clock frequency in Hertz.
+                        Default value is 10 MHz.
+        """
+        if source in CLK_SRC_MAP.keys():
+            self.set_attribute(NIRFSG_ATTR_REF_CLOCK_SOURCE,
+                               CLK_SRC_MAP[source])
+        else:
+            warnings.warn(f"Unknown clock source {source}."
+                          f" Valid sources are {CLK_SRC_MAP.keys()}. Falling"
+                          f" back to onboard clock.", RuntimeWarning)
+            self.set_attribute(NIRFSG_ATTR_REF_CLOCK_SOURCE,
+                               CLK_SRC_MAP['onboard'])
```

Comment:
is it desired to have a fallback for this instrument? in case it's not, then the `else` part can be removed, and `CLK_SRC_MAP` can be used in the parameter's validator (via `val_mapping`), and extra information can be added via `docstring` kwarg of the parameter.